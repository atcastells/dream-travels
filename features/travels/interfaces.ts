export interface Travel {
    id:          number;
    title:       string;
    description: string;
    photo_url:   string;
    status:      TravelStatus;
    itinerary:   Itinerary[];
}

export interface Itinerary {
    day:         number;
    location:    string;
    description: string;
}

export enum TravelStatus {
    TODO = "todo",
    DONE = "done",
}
