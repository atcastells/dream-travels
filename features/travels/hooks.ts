import {
  QueryKey,
  UndefinedInitialDataOptions,
  useMutation,
  useQuery,
  useQueryClient,
} from "@tanstack/react-query";
import { Travel, TravelCreation } from "./interfaces";
import {
  createTravel,
  deleteTravel,
  getTravels,
  updateTravel,
} from "./service";

export const TRAVEL_KEY = "travels";

export const useTravels = (
  travelFilters?: Partial<Travel>,
  options?: UndefinedInitialDataOptions<Travel[], Error, Travel[], QueryKey>
) => {
  const { data, ...rest } = useQuery<Travel[]>({
    ...options,
    queryKey: [TRAVEL_KEY, { travelFilters }],
    queryFn: () => getTravels(travelFilters),
  });

  return { travels: data, ...rest };
};

export const useCreateTravel = () => {
  const queryClient = useQueryClient();
  const { mutate, mutateAsync } = useMutation({
    mutationFn: (travel: TravelCreation) => {
      return createTravel(travel);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: [TRAVEL_KEY],
      });
    },
  });

  return {
    createTravel: mutate,
    createTravelAsync: mutateAsync,
  };
};

export const useDeleteTravel = () => {
  const queryClient = useQueryClient();
  const { mutate, mutateAsync } = useMutation({
    mutationFn: deleteTravel,
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: [TRAVEL_KEY],
      });
    },
  });

  return {
    deleteTravel: mutate,
    deleteTravelAsync: mutateAsync,
  };
};

export const useUpdateTravel = () => {
  const queryClient = useQueryClient();
  const { mutate, mutateAsync } = useMutation({
    mutationFn: updateTravel,
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: [TRAVEL_KEY],
      });
    },
  });

  return {
    updateTravel: mutate,
    updateTravelAsync: mutateAsync,
  };
};
