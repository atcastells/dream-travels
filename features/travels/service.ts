import axios from "axios";
import { Travel, TravelCreation } from "./interfaces";
// You can use GET, POST, PUT, PATCH and DELETE. Changes aren't persisted between calls.

export const apiInstance = axios.create({
  baseURL: "https://my-json-server.typicode.com/mariosanz92/dream-travels-data",
});

export const getTravels = async (
  travelFilters?: Partial<Travel>
): Promise<Travel[]> => {
  const filterParams = new URLSearchParams();

  if (travelFilters) {
    Object.entries(travelFilters).forEach(([key, value]) => {
      if (value) {
        filterParams.append(key, value as string);
      }
    });
  }

  const response = await apiInstance.get("/travels", {
    params: filterParams,
  });
  return response.data;
};

export const getTravel = async (id: number): Promise<Travel> => {
  const response = await apiInstance.get(`/travels/${id}`);
  return response.data;
};

export const createTravel = async (travel: TravelCreation): Promise<Travel> => {
  const response = await apiInstance.post("/travels", travel);
  return response.data;
};

export const updateTravel = async (travel: Travel): Promise<Travel> => {
  const response = await apiInstance.patch(`/travels/${travel.id}`, travel);
  return response.data;
};

export const deleteTravel = async (id: string): Promise<void> => {
  await apiInstance.delete(`/travels/${id}`);
};
