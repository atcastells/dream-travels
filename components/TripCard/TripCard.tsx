import { Travel } from "@/features/travels/interfaces";

interface TripCardProps {
  onEdit: (travelId: string) => void;
  onDelete: (travelId: string) => void;
  onDetail: (travelId: string) => void;
  travel: Travel;
}

export const TripCard = ({
  travel,
  onDelete,
  onDetail,
  onEdit,
}: TripCardProps) => {
  return (
    <div className="flex flex-row h-60 bg-gray-100 rounded-r-lg">
      <div className="w-3/4 md:w-1/2 relative">
        <img
          src={travel.photo_url}
          alt={travel.title}
          className="object-cover h-full w-full rounded-l-lg"
        />
      </div>
      <div className="flex flex-col w-3/4 md:w-1/2 p-4">
        <h1 className="text-3xl mb-8">{travel.title}</h1>
        <p className="text-lg mb-4 line-clamp-3">
          Classic tour of Portugal's vibrant cities and cultural heritage,
          including Lisbon, Porto, Fatima and the flamboyant architecture of
          Sintra.
        </p>
        <div className="flex flex-row justify-between mb-4">
          <span
            onClick={() => onDetail(travel.id)}
            className="underline cursor-pointer"
          >
            See trip detail
          </span>
          <div className="flex gap-4">
            <span
              onClick={() => onEdit(travel.id)}
              className="underline cursor-pointer"
            >
              Edit
            </span>
            <span
              onClick={() => onDelete(travel.id)}
              className="underline text-red-400 cursor-pointer"
            >
              Delete
            </span>
          </div>
        </div>
      </div>
    </div>
  );
};
