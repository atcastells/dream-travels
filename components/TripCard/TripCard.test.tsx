import { expect, test } from "@jest/globals";

import { render, screen } from "@testing-library/react";

import { TravelStatus } from "@/features/travels/interfaces";
import { TripCard } from "./TripCard";

test("renders TripCard component", () => {
  const { baseElement, findByText } = render(
    <TripCard
      travel={{
        id: "1",
        title: "Trip to Portugal",
        description:
          "Classic tour of Portugal's vibrant cities and cultural heritage, including Lisbon, Porto, Fatima and the flamboyant architecture of Sintra.",
        photo_url: "https://source.unsplash.com/1600x900/?portugal",
        status: TravelStatus.TODO,
        itinerary: [
          {
            day: 1,
            location: "Lisbon",
            description:
              "Visit the historic district of Belem and the Jeronimos Monastery.",
          },
          {
            day: 2,
            location: "Porto",
            description:
              "Explore the city's historic center and the Douro River.",
          },
          {
            day: 3,
            location: "Fatima",
            description: "Visit the Sanctuary of Our Lady of Fatima.",
          },
          {
            day: 4,
            location: "Sintra",
            description:
              "Discover the flamboyant architecture of the Pena Palace and the Moorish Castle.",
          },
        ],
      }}
      onDelete={() => {}}
      onDetail={() => {}}
      onEdit={() => {}}
    />
  );

  expect(screen.getByText("Trip to Portugal")).toBeTruthy();
  expect(
    screen.getByText(
      "Classic tour of Portugal's vibrant cities and cultural heritage, including Lisbon, Porto, Fatima and the flamboyant architecture of Sintra."
    )
  ).toBeTruthy();
  expect(screen.getByText("See trip detail")).toBeTruthy();
  expect(screen.getByText("Edit")).toBeTruthy();
  expect(screen.getByText("Delete")).toBeTruthy();

  expect(findByText("Lisbon")).toBeTruthy();

  expect(findByText("Porto")).toBeTruthy();

  // snapshot testing

  expect(baseElement).toMatchSnapshot();
});
