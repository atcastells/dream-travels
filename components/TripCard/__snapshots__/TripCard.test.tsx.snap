// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TripCard component 1`] = `
<body>
  <div>
    <div
      class="flex flex-row h-60 bg-gray-100 rounded-r-lg"
    >
      <div
        class="w-3/4 md:w-1/2 relative"
      >
        <img
          alt="Trip to Portugal"
          class="object-cover h-full w-full rounded-l-lg"
          src="https://source.unsplash.com/1600x900/?portugal"
        />
      </div>
      <div
        class="flex flex-col w-3/4 md:w-1/2 p-4"
      >
        <h1
          class="text-3xl mb-8"
        >
          Trip to Portugal
        </h1>
        <p
          class="text-lg mb-4 line-clamp-3"
        >
          Classic tour of Portugal's vibrant cities and cultural heritage, including Lisbon, Porto, Fatima and the flamboyant architecture of Sintra.
        </p>
        <div
          class="flex flex-row justify-between mb-4"
        >
          <span
            class="underline cursor-pointer"
          >
            See trip detail
          </span>
          <div
            class="flex gap-4"
          >
            <span
              class="underline cursor-pointer"
            >
              Edit
            </span>
            <span
              class="underline text-red-400 cursor-pointer"
            >
              Delete
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
